contracts/l2/FuseController.sol

```diff
@@ -0,0 +1,309 @@
+        if (address(upgradeContract) == address(0)) {
+            revert CannotUpgrade();
+        }
+
+        // Unpack the tokenData of the node.
+        bytes memory tokenData = registry.getData(uint256(node));
+        (
+            address owner,
+            address resolver,
+            uint64 expiry,
+            uint96 fuses,
+            address renewalController
+        ) = _unpack(tokenData);
+
+        bool isAuthorized = registry.getAuthorization(
+            uint256(node),
+            owner,
+            msg.sender
+        );
+
+        if (owner != msg.sender && !isAuthorized) {
+            revert Unauthorised(node, msg.sender);
+        }
+
+        if (_isExpired(tokenData)) {
+            revert nameExpired(node);
+        }
+
+        // Change the controller to the upgrade contract.
+        registry.setNode(
+            uint256(node),
+            _pack(
+                address(upgradeContract),
+                owner,
+                resolver,
+                expiry,
+                fuses,
+                renewalController
+            )
+        );
+
+        // Call the new contract to notify it of the upgrade.
+        upgradeContract.upgradeFrom(node, extraData);
+    }
+
+    /*******************
+     * Node Owner functions *
+     *******************/
+
+    function setResolver(uint256 id, address newResolver) external {
+        // get tokenData
+        bytes memory tokenData = registry.getData(id);
+        (
+            address owner,
+            ,
+            uint64 expiry,
+            uint96 fuses,
+            address renewalController
+        ) = _unpack(tokenData);
+        bool isAuthorized = registry.getAuthorization(id, owner, msg.sender);
+
+        if (owner != msg.sender && !isAuthorized) {
+            revert Unauthorised(bytes32(id), msg.sender);
+        }
+
+        registry.setNode(
+            id,
+            _pack(
+                address(this),
+                owner,
+                newResolver,
+                expiry,
+                fuses,
+                renewalController
+            )
+        );
+    }
+
+    function setSubnode(
+        bytes32 node,
+        uint256 label,
+        address subnodeOwner,
+        address subnodeResolver,
+        uint64 subnodeExpiry,
+        uint96 subnodeFuses,
+        address subnodeRenewalController
+    ) external {
+        bytes memory tokenData = registry.getData(uint256(node));
+        (address owner, , , , ) = _unpack(tokenData);
+        bool isAuthorized = registry.getAuthorization(
+            uint256(node),
+            owner,
+            msg.sender
+        );
+
+        if (owner != msg.sender && !isAuthorized) {
+            revert Unauthorised(node, msg.sender);
+        }
+
+        registry.setSubnode(
+            uint256(node),
+            label,
+            _pack(
+                address(this),
+                subnodeOwner,
+                subnodeResolver,
+                subnodeExpiry,
```

Comment:
Shouldn't there be some enforcement here that a subnode can't expire after its parent?